serve_index.py

```diff
@@ -0,0 +1,160 @@
+        # Sort by time and then by ID in case of equivalent times
+        sort_trade_key = lambda trade: (self._get_trade_time(trade), int(trade["id"]))
+        return sorted(trades, key=sort_trade_key)
+
+    def _store_new_trades(self, trades):
+        for trade in trades:
+            if (
+                self.last_stored_trade is None or
+                (self._get_trade_time(trade), int(trade["id"])) >
+                (self._get_trade_time(self.last_stored_trade), int(self.last_stored_trade["id"]))
+            ):
+                self.latest_trades_by_market[self.market] = trade
+                self.last_stored_trade = trade
+
+    def _new_ws_client(self):
+        return FtxWebsocketClient()
+
+    def start(self):
+        ws_client = FtxWebsocketClient()
+        while True:
+            if ws_client.lost_connection:
+                ws_client = FtxWebsocketClient()
+            self.ws_trades = self._sort_trades(self._get_ws_trades(ws_client))
+            self._store_new_trades(self.ws_trades)
+            time.sleep(.1)
+
+
+def stream_trades(market, latest_trades_by_market):
+    streamer = TradeUpdateStreamer(market, latest_trades_by_market)
+    streamer.start()
+
+
+def get_weighted_index_price(latest_trades_by_market, markets, weights):
+    ONE_BP = 0.0001
+    prices_ordered_by_market = [latest_trades_by_market[market]["price"] for market in markets]
+    median_price = statistics.median(prices_ordered_by_market)
+    min_price_for_index_average = median_price - (30 * ONE_BP * median_price)
+    max_price_for_index_average = median_price + (30 * ONE_BP * median_price)
+    prices_for_index_average = [
+        # Cap each element of the weighted average at 30 basis points from the median
+        min(max(price, min_price_for_index_average), max_price_for_index_average)
+        for price in prices_ordered_by_market
+    ]
+    index_price = 0
+    for price, weight in zip(prices_for_index_average, weights):
+        index_price += (price * weight)
+
+    return str(index_price)
+
+    # For testing
+    # return {
+    #     "index_price": str(index_price),
+    #     "prices_for_avg": prices_for_index_average,
+    #     "median_price": median_price,
+    #     "min_price_for_index_average": min_price_for_index_average,
+    #     "max_price_for_index_average": max_price_for_index_average,
+    #     "times": [latest_trades_by_market[market]["time"] for market in markets],
+    # }
+
+
+if __name__ == "__main__":
+    if len(sys.argv) < 2:
+        raise ValueError("Must specify the path to an input file containing all markets to track")
+
+    markets_path = sys.argv[1]
+    with open(markets_path, "r") as f:
+        configured_markets = set([line.rstrip("\n") for line in f.readlines()])
+
+    latest_trades_by_market = Manager().dict()
+    for market in configured_markets:
+        stream_proc = Process(target=stream_trades, args=(market, latest_trades_by_market,))
+        stream_proc.start()
```

Comment:
Create a multiprocess dictionary and fork a process that streams trades to each market, updating the dictionary